### ISSUE TITLE
Firefly项目的比ChatGLM2官方更加充分高效的多轮对话训练方法，是否会遗漏信息？

### ISSUE THREAD

**Issue by mvllwong** [yangjianxin1/Firefly#99]
多轮对话训练方法在做并行预测的时候：
1.是否会遗失之前对话里面的target回复信息？https://github.com/yangjianxin1/Firefly/blob/cccd542e054abfb08dfbaeec8bbc327c2f19e1ee/component/collator.py#L27-L36
2.是否会用到后向数据的信息，变成了**双向填空预测**，而不是**单向解码生成**？
新加的attention_mask和glm自制的attention mask有逻辑冲突，详见https://huggingface.co/THUDM/chatglm2-6b/blob/0ade0d38ac00258ae09450696315c2ff0b1faf12/modeling_chatglm.py#L627 的**判断条件**

**Comment by mvllwong** [comment]
有没有试验论证这种类似bert的mask方式用于生成式训练时候，对于大模型学习生成答案的自然语言逻辑顺序上的影响？

**Comment by ridiculouz** [comment]
casual lm会带有casual mask，计算前面token的时候应该看不到后面的信息吧。

**Comment by mvllwong** [comment]
> 

不知道你说的是哪个“自带casual mask”，这个firefly自己改掉了原版“casual mask”的attention_mask函数，如果你找到这两个mask函数有组合或者兼并的地方，感谢你可以指出来

**Comment by mvllwong** [comment]
> casual lm会带有casual mask，计算前面token的时候应该看不到后面的信息吧。

https://github.com/yangjianxin1/Firefly/blob/cccd542e054abfb08dfbaeec8bbc327c2f19e1ee/component/collator.py#L27-L36 可以看下这个collactor改造

**Comment by mvllwong** [comment]
> casual lm会带有casual mask，计算前面token的时候应该看不到后面的信息吧。

不知道你说的是哪个“自带casual mask”，这个firefly自己改掉了原版“casual mask”的attention_mask函数，如果你找到这两个mask函数有组合或者兼并的地方，感谢你可以指出来

**Comment by Haskely** [comment]
> casual lm会带有casual mask，计算前面token的时候应该看不到后面的信息吧。

同样有疑问：https://github.com/yangjianxin1/Firefly/issues/91

**Comment by mvllwong** [comment]
> > casual lm会带有casual mask，计算前面token的时候应该看不到后面的信息吧。
> 
> 同样有疑问：#91

等回答吧，可能打开了生成式模型训练按完形填空做的新思路

**Comment by mvllwong** [comment]
> > casual lm会带有casual mask，计算前面token的时候应该看不到后面的信息吧。
> 
> 同样有疑问：#91

感谢大佬指出，这里自己设计attention mask会导致modelling代码里面默认关闭 causual attetion，详见：https://huggingface.co/THUDM/chatglm2-6b/blob/0ade0d38ac00258ae09450696315c2ff0b1faf12/modeling_chatglm.py#L192

**Comment by Haskely** [comment]
更新
chatglm2 在这里添加了 causal attention mask：https://huggingface.co/THUDM/chatglm2-6b/blob/0ade0d38ac00258ae09450696315c2ff0b1faf12/modeling_chatglm.py#L627

所以不存在该问题。

**Comment by mvllwong** [comment]
> 更新 chatglm2 在这里添加了 causal attention mask：https://huggingface.co/THUDM/chatglm2-6b/blob/0ade0d38ac00258ae09450696315c2ff0b1faf12/modeling_chatglm.py#L627
> 
> 所以不存在该问题。

我想我在你的#91号问题里已经和你讲过这个get_masks函数了，他同样是if attention_mask is None才会触发的，如果不执行这个函数，就会导致”这个问题不仅如此，事实上gllm自制的attention mask维度和firefly自以为的维度是不同的，甚至和prefix encode mask叠加时候都可能有问题“的一系列问题

**Comment by Haskely** [comment]
> 我想我在你的#91号问题里已经和你讲过这个get_masks函数了，他同样是if attention_mask is None才会触发的
1. #91 没看到有提到过 get_masks 函数。
2. get_masks 是在 if attention_mask is **not** None 且不为全 1 时触发的，见 https://huggingface.co/THUDM/chatglm2-6b/blob/v1.0/modeling_chatglm.py#L729；而且若 attention_mask 为全 1，直接会忽略掉，full_attention_mask 保持为 None，此时相当于没传 attention_mask。因此逻辑上的确没问题。

> ”这个问题不仅如此，事实上gllm自制的attention mask维度和firefly自以为的维度是不同的，甚至和prefix encode mask叠加时候都可能有问题“
描述不详，没有对应给出代码对比以及详细解释。

**Comment by mvllwong** [comment]
> > 我想我在你的#91号问题里已经和你讲过这个get_masks函数了，他同样是if attention_mask is None才会触发的
> 
> 1. [Chatglm2-6b ~~不能~~ 可以直接传入 attention_mask #91](https://github.com/yangjianxin1/Firefly/issues/91) 没看到有提到过 get_masks 函数。
> 2. get_masks 是在 if attention_mask is **not** None 且不为全 1 时触发的，见 https://huggingface.co/THUDM/chatglm2-6b/blob/v1.0/modeling_chatglm.py#L729 ；而且若 attention_mask 为全 1，直接会忽略掉，full_attention_mask 保持为 None，此时相当于没传 attention_mask。因此逻辑上的确没问题。
> 
> > ”这个问题不仅如此，事实上gllm自制的attention mask维度和firefly自以为的维度是不同的，甚至和prefix encode mask叠加时候都可能有问题“
> 
> 描述不详，没有对应给出代码对比以及详细解释。

感谢指出问题，下面的回复我都会附上解释的对应代码，避免信息不详或者难以同步。
1. if attention_mask is None和你看到的不同的问题：是我的问题，我看的代码是老代码，参见：https://huggingface.co/THUDM/chatglm-6b/blob/08bc85104db4e8da2c215a29c469218953056251/modeling_chatglm.py#L916

**Comment by mvllwong** [comment]
> > 我想我在你的#91号问题里已经和你讲过这个get_masks函数了，他同样是if attention_mask is None才会触发的
> 
> 1. [Chatglm2-6b ~~不能~~ 可以直接传入 attention_mask #91](https://github.com/yangjianxin1/Firefly/issues/91) 没看到有提到过 get_masks 函数。
> 2. get_masks 是在 if attention_mask is **not** None 且不为全 1 时触发的，见 https://huggingface.co/THUDM/chatglm2-6b/blob/v1.0/modeling_chatglm.py#L729 ；而且若 attention_mask 为全 1，直接会忽略掉，full_attention_mask 保持为 None，此时相当于没传 attention_mask。因此逻辑上的确没问题。
> 
> > ”这个问题不仅如此，事实上gllm自制的attention mask维度和firefly自以为的维度是不同的，甚至和prefix encode mask叠加时候都可能有问题“
> 
> 描述不详，没有对应给出代码对比以及详细解释。


感谢指出问题，我的提问里确实缺少了代码行作为信息补充，导致了信息不详和不同步的问题，下面我会注意：

**Comment by mvllwong** [comment]
> > 我想我在你的#91号问题里已经和你讲过这个get_masks函数了，他同样是if attention_mask is None才会触发的
> 
> 1. [Chatglm2-6b ~~不能~~ 可以直接传入 attention_mask #91](https://github.com/yangjianxin1/Firefly/issues/91) 没看到有提到过 get_masks 函数。
> 2. get_masks 是在 if attention_mask is **not** None 且不为全 1 时触发的，见 https://huggingface.co/THUDM/chatglm2-6b/blob/v1.0/modeling_chatglm.py#L729 ；而且若 attention_mask 为全 1，直接会忽略掉，full_attention_mask 保持为 None，此时相当于没传 attention_mask。因此逻辑上的确没问题。
> 
> > ”这个问题不仅如此，事实上gllm自制的attention mask维度和firefly自以为的维度是不同的，甚至和prefix encode mask叠加时候都可能有问题“
> 
> 描述不详，没有对应给出代码对比以及详细解释。

感谢指出问题，我的提问里确实缺少了代码行作为信息补充，导致了信息不详和不同步的问题，下面我会注意：
1.get_masks 的判断条件为什么我看到的是 if attention_mask is None: 因为我看的是老代码，这里是我的问题，没有同步到最新。旧版代码如下：
https://huggingface.co/THUDM/chatglm-6b/blob/08bc85104db4e8da2c215a29c469218953056251/modeling_chatglm.py#L916
https://huggingface.co/THUDM/chatglm-6b/blob/08bc85104db4e8da2c215a29c469218953056251/modeling_chatglm.py#L1118
2.”而且若 attention_mask 为全 1，直接会忽略掉，full_attention_mask 保持为 None，此时相当于没传 attention_mask。因此逻辑上的确没问题。“回到一开始的讨论，此时到底有没有问题：
”而且若 attention_mask 为全 1，直接会忽略掉，full_attention_mask 保持为 None“很明显你这里说的是get_masks在最新的代码里未被调用，也就是和你说的”更新
chatglm2 在这里添加了 causal attention mask：https://huggingface.co/THUDM/chatglm2-6b/blob/0ade0d38ac00258ae09450696315c2ff0b1faf12/modeling_chatglm.py#L627

所以不存在该问题。“形成了前后逻辑矛盾，get_maks的causual attetion 还是没加上，还是有问题

**Comment by Haskely** [comment]
> 很明显你这里说的是get_masks在最新的代码里未被调用 ... get_mask 的 causal attention 还是没加上，还是有问题

当 `get_mask` 未被调用时，意味着 `full_attention_mask` 为 None，相当于没有传 `attention_mask` (注意调用时只会传入 `full_attention_mask` 变量 https://huggingface.co/THUDM/chatglm2-6b/blob/0ade0d38ac00258ae09450696315c2ff0b1faf12/modeling_chatglm.py#L741-L742)。此时 chatglm2 会在这三行自动加上 causal mask：https://huggingface.co/THUDM/chatglm2-6b/blob/0ade0d38ac00258ae09450696315c2ff0b1faf12/modeling_chatglm.py#L191-L193 其余情况会调用 `get_mask`，则会在 `get_mask` 函数内部添加 causal mask ，所以两个分路终究都加了 causal mask。

**Comment by mvllwong** [comment]
> > 很明显你这里说的是get_masks在最新的代码里未被调用 ... get_mask 的 causal attention 还是没加上，还是有问题
> 
> 当 `get_mask` 未被调用时，意味着 `full_attention_mask` 为 None，相当于没有传 `attention_mask` (注意调用时只会传入 `full_attention_mask` 变量 https://huggingface.co/THUDM/chatglm2-6b/blob/0ade0d38ac00258ae09450696315c2ff0b1faf12/modeling_chatglm.py#L741-L742 )。此时 chatglm2 会在这三行自动加上 causal mask：https://huggingface.co/THUDM/chatglm2-6b/blob/0ade0d38ac00258ae09450696315c2ff0b1faf12/modeling_chatglm.py#L191-L193 其余情况会调用 `get_mask`，则会在 `get_mask` 函数内部添加 causal mask ，所以两个分路终究都加了 causal mask。
> 
> > 因为我看的是老代码，这里是我的问题，没有同步到最新
> 
> 我发现你给的是 chatglm 1代的代码。我只看了 2代的，对 1代没有研究过，或许真的存在问题。

你的梳理我理解了，问题在于你提出的这一行是最底层的core attention的forward 接受的attention_mask
https://huggingface.co/THUDM/chatglm2-6b/blob/0ade0d38ac00258ae09450696315c2ff0b1faf12/modeling_chatglm.py#L741-L742 ，
不对应你#91之前提到的：https://github.com/yangjianxin1/Firefly/blob/master/component/dataset.py#L96  
中输入的attention_mask（可能还有一个batch padding的步骤）。
结论是：firefly的设计要点在tartget_mask上，attention_mask可有可无，也并不涉及改变掩码矩阵等

**Comment by mvllwong** [comment]
> casual lm会带有casual mask，计算前面token的时候应该看不到后面的信息吧。

你说的对的，我理解错了，这里只是做了个losses相关的targets重选（末尾的换成所有的），和mask设计可以说没什么关系
